File: HelmNet/smarthelm/backend/firestore_reporter.py
"""
firestore_reporter.py — Publishes Pi helmet data to Firebase Firestore.

Mirrors FirestoreReporter.kt (Android) so Pi-based helmet riders appear
on the fleet dashboard alongside phone-based riders with the same schema.

Schema written to  riders/{helmet_id}:
    riderName, managerId, emergencyContact
    eyeState, perclos, continuousClosureSec, alertActive, faceDetected
    heartRate, spo2                          ← Pi-only biometric fields
    eyeLandmarksLeft, eyeLandmarksRight      ← normalized flat lists
    snapshot                                 ← base64 eye-strip JPEG (180x72)
    connected, updatedAt, smsNeeded

Setup:
    pip install firebase-admin opencv-python-headless
    Place serviceAccountKey.json in the same directory as this file.
    Set HELMET_ID and FLEET_MANAGER_PHONE in config.py.
"""

import base64
import io
import logging
import threading
import time

import cv2
import numpy as np

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Optional Firebase import — module still loads if firebase-admin is absent
# ---------------------------------------------------------------------------
try:
    import firebase_admin
    from firebase_admin import credentials, firestore
    _FIREBASE_AVAILABLE = True
except ImportError:
    _FIREBASE_AVAILABLE = False
    logger.warning("firebase-admin not installed — Firestore reporting disabled")


# ---------------------------------------------------------------------------
# Eye-strip crop constants (matches Android FirestoreReporter)
# ---------------------------------------------------------------------------
SNAPSHOT_W       = 180
SNAPSHOT_H       = 72
SNAPSHOT_QUALITY = 55      # JPEG quality %
SNAPSHOT_INTERVAL_NORMAL = 0.35   # seconds (~3 fps)
SNAPSHOT_INTERVAL_ALERT  = 0.10   # seconds (~10 fps)
STATUS_INTERVAL          = 2.0    # seconds


class FirestoreReporter:
    """
    Thread-safe Firestore publisher for Pi helmet data.
    All Firestore writes happen on a background daemon thread.
    """

    def __init__(self, helmet_id: str, rider_name: str,
                 emergency_contact: str = "", manager_id: str = "",
                 service_account_path: str = "serviceAccountKey.json"):
        self._helmet_id        = helmet_id
        self._rider_name       = rider_name
        self._emergency_contact = emergency_contact
        self._manager_id       = manager_id
        self._enabled          = False

        self._last_status_t   = 0.0
        self._last_snapshot_t = 0.0
        self._prev_alert      = False

        if not _FIREBASE_AVAILABLE:
            return

        try:
            if not firebase_admin._apps:
                cred = credentials.Certificate(service_account_path)
                firebase_admin.initialize_app(cred)
            self._db  = firestore.client()
            self._ref = self._db.collection("riders").document(helmet_id)
            self._enabled = True
            logger.info(f"FirestoreReporter: connected — helmet_id={helmet_id}")
        except Exception as e:
            logger.warning(f"FirestoreReporter: init failed ({e})")

    # ------------------------------------------------------------------
    # Public API — called from app.py inference loop
    # ------------------------------------------------------------------

    def on_frame(self, eye_state: str, perclos: float,
                 continuous_sec: float, alert_active: bool,
                 face_detected: bool, heart_rate: float = 0.0,
                 spo2: float = 0.0,
                 left_eye_norm=None, right_eye_norm=None):
        """Push status to Firestore (throttled to STATUS_INTERVAL)."""
        if not self._enabled:
            return

        now       = time.time()
        alert_edge = alert_active != self._prev_alert
        self._prev_alert = alert_active

        if not alert_edge and (now - self._last_status_t) < STATUS_INTERVAL:
            return
        self._last_status_t = now

        left_flat  = [v for pt in (left_eye_norm  or []) for v in pt]
        right_flat = [v for pt in (right_eye_norm or []) for v in pt]

        data = {
            "riderName":            self._rider_name,
            "managerId":            self._manager_id,
            "emergencyContact":     self._emergency_contact,
            "eyeState":             eye_state,
            "perclos":              round(perclos, 2),
            "continuousClosureSec": round(continuous_sec, 2),
            "alertActive":          alert_active,
            "faceDetected":         face_detected,
            "heartRate":            round(heart_rate, 1),
            "spo2":                 round(spo2, 1),
            "eyeLandmarksLeft":     left_flat,
            "eyeLandmarksRight":    right_flat,
            "connected":            True,
            "updatedAt":            firestore.SERVER_TIMESTAMP,
        }
        if alert_edge:
            data["smsNeeded"] = alert_active   # Cloud Function → MSG91 SMS

        threading.Thread(
            target=self._write, args=(data,), daemon=True, name="fs-status"
        ).start()

    def maybe_update_snapshot(self, frame: np.ndarray,
                              left_eye_norm=None, right_eye_norm=None,
                              alert_active: bool = False):
        """
        Crop to eye strip, encode as JPEG, push to Firestore.
        Frame must be a BGR numpy array (OpenCV format).
        """
        if not self._enabled:
            return

        now      = time.time()
        interval = SNAPSHOT_INTERVAL_ALERT if alert_active else SNAPSHOT_INTERVAL_NORMAL
        if (now - self._last_snapshot_t) < interval:
            return
        self._last_snapshot_t = now

        threading.Thread(
            target=self._build_and_push_snapshot,
            args=(frame.copy(), left_eye_norm, right_eye_norm, alert_active),
            daemon=True, name="fs-snapshot"
        ).start()

    def flush(self):
        """Mark rider as disconnected on shutdown."""
        if not self._enabled:
            return
        try:
            self._ref.update({
                "connected": False,
                "smsNeeded": False,
                "updatedAt": firestore.SERVER_TIMESTAMP,
            })
            logger.info("FirestoreReporter: flushed disconnect status")
        except Exception as e:
            logger.warning(f"FirestoreReporter: flush failed ({e})")

    # ------------------------------------------------------------------
    # Private — snapshot pipeline
    # ------------------------------------------------------------------

    def _build_and_push_snapshot(self, frame, left_eye, right_eye, alert_active):
        try:
            annotated = frame.copy()

            # Draw eye outlines on the full frame
            if left_eye and right_eye:
                self._draw_eye_overlay(annotated, left_eye, right_eye, alert_active)

            # Crop to eye strip using landmark bounding box
            if left_eye and right_eye:
                crop = self._crop_eye_strip(annotated, left_eye, right_eye)
            else:
                h, w = annotated.shape[:2]
                crop = annotated[h//3: 2*h//3, :]   # fallback: middle third

            # Scale to fixed output size
            scaled = cv2.resize(crop, (SNAPSHOT_W, SNAPSHOT_H), interpolation=cv2.INTER_LINEAR)

            # JPEG encode
            encode_params = [cv2.IMWRITE_JPEG_QUALITY, SNAPSHOT_QUALITY]
            _, buf = cv2.imencode(".jpg", scaled, encode_params)
            b64 = base64.b64encode(buf.tobytes()).decode("ascii")

            self._ref.update({
                "snapshot":          b64,
                "snapshotUpdatedAt": firestore.SERVER_TIMESTAMP,
            })
        except Exception as e:
            logger.debug(f"FirestoreReporter: snapshot failed ({e})")

    def _crop_eye_strip(self, frame, left_eye, right_eye):
        h, w = frame.shape[:2]
        all_pts = list(left_eye) + list(right_eye)
        xs = [p[0] for p in all_pts]; ys = [p[1] for p in all_pts]

        span_x = max(max(xs) - min(xs), 0.05)
        span_y = max(max(ys) - min(ys), 0.015)
        pad_x  = span_x * 0.50
        pad_y  = span_y * 2.50

        left   = max(0,   int((min(xs) - pad_x) * w))
        top    = max(0,   int((min(ys) - pad_y) * h))
        right  = min(w,   int((max(xs) + pad_x) * w))
        bottom = min(h,   int((max(ys) + pad_y) * h))

        if right <= left or bottom <= top:
            return frame
        return frame[top:bottom, left:right]

    def _draw_eye_overlay(self, frame, left_eye, right_eye, alert_active):
        h, w = frame.shape[:2]
        color = (34, 204, 68) if not alert_active else (68, 68, 255)  # BGR

        for eye in (left_eye, right_eye):
            pts = np.array([[int(x*w), int(y*h)] for x, y in eye], np.int32)
            cv2.polylines(frame, [pts], isClosed=True, color=color, thickness=2)
            for px, py in pts:
                cv2.circle(frame, (px, py), 3, color, -1)

        if alert_active:
            overlay = frame.copy()
            cv2.rectangle(overlay, (0, 0), (w, h), (0, 0, 180), -1)
            cv2.addWeighted(overlay, 0.15, frame, 0.85, 0, frame)

    def _write(self, data: dict):
        try:
            self._ref.set(data, merge=True)
        except Exception as e:
            logger.warning(f"FirestoreReporter: write failed ({e})")

File: HelmNet/mobile/android/app/src/main/java/com/smarthelm/mobile/cloud/FirestoreReporter.kt
package com.smarthelm.mobile.cloud

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.Path
import android.os.SystemClock
import android.util.Base64
import android.util.Log
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.GeoPoint
import com.google.firebase.firestore.SetOptions
import com.smarthelm.mobile.detection.DetectionResult
import com.smarthelm.mobile.detection.FatigueResult
import com.smarthelm.mobile.detection.PerclosResult
import java.io.ByteArrayOutputStream

/**
 * Publishes rider status + annotated eye-region frames to Firestore.
 *
 * Video pipeline (privacy + clarity):
 *   1. Rotate full frame by imageProxy.rotationDegrees → upright face
 *   2. Transform landmark coords into the rotated frame's coordinate space
 *   3. Draw MediaPipe eye outlines + EAR lines on the upright frame
 *   4. Crop ONLY the eye strip (both eyes + padding) — no face, no body
 *   5. Scale to 180×72 at 55% JPEG quality → ~5–10 KB, crisp and private
 *   6. Push to Firestore at 350 ms normal / 100 ms alert
 */
class FirestoreReporter(
    private val db: FirebaseFirestore,
    private val deviceId: String,
    private val riderName: String,
    private val emergencyContact: String = "",
    private val managerId: String = ""
) {

    companion object {
        private const val TAG                = "FirestoreReporter"
        private const val THROTTLE_MS        = 2_000L
        private const val SNAPSHOT_NORMAL_MS = 350L    // ~3 fps
        private const val SNAPSHOT_ALERT_MS  = 100L    // ~10 fps
        private const val SNAPSHOT_W         = 180     // eye strip output width
        private const val SNAPSHOT_H         = 72      // eye strip output height
        private const val SNAPSHOT_QUALITY   = 55
        private const val NO_FACE_W          = 80      // fallback when no face detected
        private const val NO_FACE_H          = 60
        private const val NO_FACE_QUALITY    = 28
    }

    private var lastPushMs      = 0L
    private var lastSnapshotMs  = 0L
    private var prevAlertActive = false
    private var lastLocation: GeoPoint? = null
    private var lastSpeedKmph: Float = -1f   // -1 = unknown
    private var lastBearing: Float   = -1f

    // Trip / liveness state (F2 — app-off-mid-trip safety concern)
    private var tripId: String? = null
    private var tripActive = false
    private var appState = "FOREGROUND"      // FOREGROUND | BACKGROUND | ENDED

    // Reusable paints — allocated once
    private val eyePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.STROKE; strokeWidth = 2.8f }
    private val dotPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.FILL }
    private val earPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        color = Color.rgb(255, 220, 0); strokeWidth = 1.6f; style = Paint.Style.STROKE
    }
    private val tintPaint = Paint()

    fun setLastLocation(lat: Double, lng: Double, speedKmph: Float = -1f, bearing: Float = -1f) {
        lastLocation  = GeoPoint(lat, lng)
        lastSpeedKmph = speedKmph
        lastBearing   = bearing
    }

    // ── Trip / heartbeat (F2) ──────────────────────────────────────────────────
    // The dashboard distinguishes a graceful trip end (appState=ENDED / connected=false)
    // from a rider going dark mid-trip (tripActive=true, appState!=ENDED, heartbeat stale).

    fun setAppState(state: String) { appState = state }

    fun startTrip(id: String) {
        tripId = id
        tripActive = true
        pushTrip(mapOf(
            "tripActive"     to true,
            "tripId"         to id,
            "tripStartedAt"  to FieldValue.serverTimestamp(),
            "appState"       to appState
        ))
    }

    /** Lightweight liveness ping — keeps lastHeartbeatAt fresh even when frames stall. */
    fun heartbeat() {
        if (!tripActive) return
        pushTrip(mapOf("tripActive" to true, "appState" to appState))
    }

    /** Graceful trip end — must NOT raise a safety concern on the dashboard. */
    fun endTrip() {
        tripActive = false
        appState = "ENDED"
        pushTrip(mapOf("tripActive" to false, "appState" to "ENDED"))
    }

    private fun pushTrip(extra: Map<String, Any?>) {
        val doc = HashMap<String, Any?>(extra)
        doc["connected"]       = true
        doc["lastHeartbeatAt"] = FieldValue.serverTimestamp()
        doc["updatedAt"]       = FieldValue.serverTimestamp()
        db.collection("riders").document(deviceId)
            .set(doc, SetOptions.merge())
            .addOnFailureListener { e -> Log.w(TAG, "Trip push failed: ${e.message}") }
    }

    // ── Status push ──────────────────────────────────────────────────────────

    /**
     * @param fatigue the fused [FatigueResult]; when present its [FatigueResult.alertActive] is the
     *   authoritative alert (already debounced + speed-gated), so the dashboard and the on-device
     *   alarm agree. Falls back to [PerclosResult.alertActive] for older callers.
     */
    fun onFrame(det: DetectionResult, perc: PerclosResult, fatigue: FatigueResult? = null) {
        val now            = SystemClock.elapsedRealtime()
        val effectiveAlert = fatigue?.alertActive ?: perc.alertActive
        val alertEdge      = effectiveAlert != prevAlertActive
        prevAlertActive    = effectiveAlert

        if (!alertEdge && (now - lastPushMs) < THROTTLE_MS) return
        lastPushMs = now

        val leftFlat  = det.leftEyeNorm.flatMap  { listOf(it.first, it.second) }
        val rightFlat = det.rightEyeNorm.flatMap { listOf(it.first, it.second) }
        val alertType = fatigue?.cause?.ifBlank { perc.alertType } ?: perc.alertType

        val status = hashMapOf<String, Any?>(
            "riderName"            to riderName,
            "managerId"            to managerId,
            "emergencyContact"     to emergencyContact,
            "eyeState"             to det.eyeState,
            "perclos"              to perc.perclos,
            "continuousClosureSec" to perc.continuousClosureSec,
            "alertActive"          to effectiveAlert,
            "alertType"            to alertType,
            "faceDetected"         to det.faceDetected,
            "eyeLandmarksLeft"     to leftFlat,
            "eyeLandmarksRight"    to rightFlat,
            "location"             to lastLocation,
            "tripActive"           to tripActive,
            "appState"             to appState,
            "lastHeartbeatAt"      to FieldValue.serverTimestamp(),
            "connected"            to true,
            "updatedAt"            to FieldValue.serverTimestamp()
        )
        tripId?.let { status["tripId"] = it }
        if (lastSpeedKmph >= 0f) status["speedKmph"] = lastSpeedKmph
        if (lastBearing   >= 0f) status["heading"]   = lastBearing
        if (fatigue != null) {
            status["fatigueScore"]     = fatigue.score
            status["fatigueLevel"]     = fatigue.level
            status["fatigueBreakdown"] = fatigue.breakdown
            status["yawnCount"]        = fatigue.yawnCount
            status["blinkRate"]        = fatigue.blinkRate
        }
        if (alertEdge) status["smsNeeded"] = effectiveAlert

        db.collection("riders").document(deviceId)
            .set(status, SetOptions.merge())
            .addOnFailureListener { e -> Log.w(TAG, "Status push failed: ${e.message}") }

        if (alertEdge && effectiveAlert) {
            val alertDoc = hashMapOf<String, Any?>(
                "type"          to alertType.ifBlank { if (perc.alertContinuous) "CONTINUOUS_CLOSURE" else "PERCLOS" },
                "perclos"       to perc.perclos,
                "continuousSec" to perc.continuousClosureSec,
                "fatigueScore"  to (fatigue?.score ?: 0f),
                "location"      to lastLocation,
                "timestamp"     to FieldValue.serverTimestamp()
            )
            db.collection("riders").document(deviceId).collection("alerts").add(alertDoc)
        }
    }

    // ── Snapshot pipeline ────────────────────────────────────────────────────

    fun maybeUpdateSnapshot(
        bitmap: Bitmap,
        rotationDegrees: Int,
        alertActive: Boolean,
        detection: DetectionResult
    ) {
        val now      = SystemClock.elapsedRealtime()
        val interval = if (alertActive) SNAPSHOT_ALERT_MS else SNAPSHOT_NORMAL_MS
        if ((now - lastSnapshotMs) < interval) return
        lastSnapshotMs = now

        try {
            // ── Step 1: Rotate to correct display orientation ────────────────
            val rotated = rotateFrame(bitmap, rotationDegrees)

            // ── Step 2: Transform landmark coords into rotated-frame space ───
            val leftRot  = transformLandmarks(detection.leftEyeNorm,  rotationDegrees)
            val rightRot = transformLandmarks(detection.rightEyeNorm, rotationDegrees)

            // ── Step 3: Draw inference overlay on upright frame ──────────────
            val annotated = rotated.copy(Bitmap.Config.ARGB_8888, true)
            rotated.recycle()
            if (detection.faceDetected && leftRot.size >= 6) {
                drawOverlay(annotated, leftRot, rightRot, detection.eyeState, alertActive)
            }

            // ── Step 4: Crop to eye-only strip (privacy) ─────────────────────
            val (outW, outH, quality) = if (detection.faceDetected && leftRot.size >= 6) {
                Triple(SNAPSHOT_W, SNAPSHOT_H, SNAPSHOT_QUALITY)
            } else {
                Triple(NO_FACE_W, NO_FACE_H, NO_FACE_QUALITY)
            }

            val cropped = if (detection.faceDetected && leftRot.size >= 6) {
                cropEyeStrip(annotated, leftRot, rightRot)
            } else {
                null
            }
            val source = cropped ?: annotated

            // ── Step 5: Scale + encode ───────────────────────────────────────
            val scaled = Bitmap.createScaledBitmap(source, outW, outH, true)
            if (cropped != null) { cropped.recycle(); annotated.recycle() }
            else                 { annotated.recycle() }

            val bos = ByteArrayOutputStream()
            scaled.compress(Bitmap.CompressFormat.JPEG, quality, bos)
            scaled.recycle()
            val b64 = Base64.encodeToString(bos.toByteArray(), Base64.NO_WRAP)

            // ── Step 6: Push ─────────────────────────────────────────────────
            db.collection("riders").document(deviceId)
                .update(
                    "snapshot",          b64,
                    "snapshotUpdatedAt", FieldValue.serverTimestamp(),
                    "updatedAt",         FieldValue.serverTimestamp()
                )
                .addOnFailureListener { e -> Log.w(TAG, "Snapshot push failed: ${e.message}") }

        } catch (e: Exception) {
            Log.w(TAG, "Snapshot pipeline failed: ${e.message}")
        }
    }

    fun flush() {
        db.collection("riders").document(deviceId).update(
            mapOf("connected" to false, "smsNeeded" to false, "updatedAt" to FieldValue.serverTimestamp())
        )
    }

    // ── Rotation ─────────────────────────────────────────────────────────────

    private fun rotateFrame(bmp: Bitmap, deg: Int): Bitmap {
        if (deg == 0) return bmp.copy(bmp.config ?: Bitmap.Config.ARGB_8888, false)
        val m = Matrix().apply { postRotate(deg.toFloat()) }
        return Bitmap.createBitmap(bmp, 0, 0, bmp.width, bmp.height, m, true)
    }

    /**
     * Transform normalized (0–1) landmark coords from original-frame space
     * into the rotated-frame space. Required before cropping from the rotated bitmap.
     *
     *  0°   (nx, ny) → (nx, ny)
     *  90°  (nx, ny) → (1-ny, nx)      [90° CW]
     * 180°  (nx, ny) → (1-nx, 1-ny)
     * 270°  (nx, ny) → (ny,   1-nx)    [270° CW = 90° CCW]
     */
    private fun transformLandmarks(
        pts: List<Pair<Float, Float>>,
        deg: Int
    ): List<Pair<Float, Float>> = pts.map { (nx, ny) ->
        when (deg % 360) {
            90  -> Pair(1f - ny, nx)
            180 -> Pair(1f - nx, 1f - ny)
            270 -> Pair(ny, 1f - nx)
            else -> Pair(nx, ny)
        }
    }

    // ── Eye strip crop ────────────────────────────────────────────────────────

    private fun cropEyeStrip(
        bmp: Bitmap,
        leftEye: List<Pair<Float, Float>>,
        rightEye: List<Pair<Float, Float>>
    ): Bitmap {
        val all = leftEye + rightEye
        val bW  = bmp.width.toFloat()
        val bH  = bmp.height.toFloat()

        val minX = all.minOf { it.first }
        val maxX = all.maxOf { it.first }
        val minY = all.minOf { it.second }
        val maxY = all.maxOf { it.second }

        val spanX = (maxX - minX).coerceAtLeast(0.05f)
        val spanY = (maxY - minY).coerceAtLeast(0.015f)

        // Horizontal: 50% padding each side (capture full eye width + brow)
        // Vertical: 250% padding (eyes are tiny vertical strip; need forehead + cheek context)
        val padX = spanX * 0.50f
        val padY = spanY * 2.50f

        val left   = ((minX - padX) * bW).coerceIn(0f, bW - 1f).toInt()
        val top    = ((minY - padY) * bH).coerceIn(0f, bH - 1f).toInt()
        val right  = ((maxX + padX) * bW).coerceIn(left.toFloat() + 1f, bW).toInt()
        val bottom = ((maxY + padY) * bH).coerceIn(top.toFloat() + 1f, bH).toInt()

        return Bitmap.createBitmap(bmp, left, top, right - left, bottom - top)
    }

    // ── Inference overlay ─────────────────────────────────────────────────────

    private fun drawOverlay(
        bmp: Bitmap,
        leftEye: List<Pair<Float, Float>>,
        rightEye: List<Pair<Float, Float>>,
        eyeState: String,
        alertActive: Boolean
    ) {
        val canvas = Canvas(bmp)
        val W = bmp.width.toFloat(); val H = bmp.height.toFloat()

        val color = when (eyeState) {
            "OPEN"   -> Color.rgb(34, 204, 68)
            "CLOSED" -> Color.rgb(255, 68,  68)
            else     -> Color.rgb(120, 120, 120)
        }
        eyePaint.color = color; dotPaint.color = color

        listOf(leftEye, rightEye).forEach { pts ->
            if (pts.size < 6) return@forEach
            val path = Path()
            pts.forEachIndexed { i, (nx, ny) ->
                val px = nx * W; val py = ny * H
                if (i == 0) path.moveTo(px, py) else path.lineTo(px, py)
            }
            path.close(); canvas.drawPath(path, eyePaint)
            pts.forEach { (nx, ny) -> canvas.drawCircle(nx * W, ny * H, 3.2f, dotPaint) }
        }

        // EAR vertical measurement lines
        listOf(leftEye, rightEye).forEach { pts ->
            if (pts.size >= 6) {
                canvas.drawLine(pts[1].first*W, pts[1].second*H, pts[5].first*W, pts[5].second*H, earPaint)
                canvas.drawLine(pts[2].first*W, pts[2].second*H, pts[4].first*W, pts[4].second*H, earPaint)
            }
        }

        if (alertActive) {
            tintPaint.color = Color.argb(40, 255, 68, 68)
            canvas.drawRect(0f, 0f, W, H, tintPaint)
        }
    }
}
